File: generators/transform.cpp

// Part of the ProceduralTextureMaker project.
// http://github.com/johanokl/ProceduralTextureMaker
// Released under GPLv3.
// Johan Lindqvist (johan.lindqvist@gmail.com)

#include "transform.h"
#include <QPainter>

TransformTextureGenerator::TransformTextureGenerator() {
   TextureGeneratorSetting scaleX;
   scaleX.name = "Horizontal scale (%)";
   scaleX.description = "Scales the input image horizontally around the texture centre.";
   scaleX.defaultvalue = QVariant((double)100);
   scaleX.min = QVariant(0);
   scaleX.max = QVariant(500);
   scaleX.group = "scale";
   scaleX.id = "xscale";
   configurables.append(scaleX);

   TextureGeneratorSetting scaleY;
   scaleY.name = "Vertical scale (%)";
   scaleY.description = "Scales the input image vertically around the texture centre.";
   scaleY.defaultvalue = QVariant((double)100);
   scaleY.min = QVariant(0);
   scaleY.max = QVariant(500);
   scaleY.group = "scale";
   scaleY.id = "yscale";
   configurables.append(scaleY);

   TextureGeneratorSetting rotation;
   rotation.name = "Rotation (°)";
   rotation.description = "Rotates the input image around the texture centre.";
   rotation.defaultvalue = QVariant((double)0);
   rotation.min = QVariant(-360);
   rotation.max = QVariant(360);
   rotation.id = "rotation";
   configurables.append(rotation);

   TextureGeneratorSetting offsetLeft;
   offsetLeft.name = "Horizontal offset (%)";
   offsetLeft.description = "Moves the input image horizontally across the texture.";
   offsetLeft.defaultvalue = QVariant((double)0);
   offsetLeft.min = QVariant(-200);
   offsetLeft.max = QVariant(200);
   offsetLeft.id = "offsetleft";
   configurables.append(offsetLeft);

   TextureGeneratorSetting offsetTop;
   offsetTop.name = "Vertical offset (%)";
   offsetTop.description = "Moves the input image vertically across the texture.";
   offsetTop.defaultvalue = QVariant((double)0);
   offsetTop.min = QVariant(-200);
   offsetTop.max = QVariant(200);
   offsetTop.id = "offsettop";
   configurables.append(offsetTop);

   TextureGeneratorSetting colorsetting;
   colorsetting.name = "Background colour";
   colorsetting.description = "Colour used where the transformed image leaves empty space.";
   colorsetting.defaultvalue = QVariant(QColor(0, 0, 0, 0));
   colorsetting.id = "backgroundcolor";
   configurables.append(colorsetting);

   TextureGeneratorSetting firstXtiles;
   firstXtiles.name = "First-pass horizontal tiles";
   firstXtiles.description = "Number of horizontal copies created before transformation.";
   firstXtiles.defaultvalue = QVariant((int)1);
   firstXtiles.min = QVariant(1);
   firstXtiles.max = QVariant(5);
   firstXtiles.group = "first tiles";
   firstXtiles.id = "firstXtiles";
   configurables.append(firstXtiles);

   TextureGeneratorSetting firstYtiles;
   firstYtiles.name = "First-pass vertical tiles";
   firstYtiles.description = "Number of vertical copies created before transformation.";
   firstYtiles.defaultvalue = QVariant((int)1);
   firstYtiles.min = QVariant(1);
   firstYtiles.max = QVariant(5);
   firstYtiles.group = "first tiles";
   firstYtiles.id = "firstYtiles";
   configurables.append(firstYtiles);

   TextureGeneratorSetting secondXtiles;
   secondXtiles.name = "Second-pass horizontal tiles";
   secondXtiles.description = "Number of horizontal copies created after transformation.";
   secondXtiles.defaultvalue = QVariant((int)1);
   secondXtiles.min = QVariant(1);
   secondXtiles.max = QVariant(50);
   secondXtiles.group = "second tiles";
   secondXtiles.id = "secondXtiles";
   configurables.append(secondXtiles);

   TextureGeneratorSetting secondYtiles;
   secondYtiles.name = "Second-pass vertical tiles";
   secondYtiles.description = "Number of vertical copies created after transformation.";
   secondYtiles.defaultvalue = QVariant((int)1);
   secondYtiles.min = QVariant(1);
   secondYtiles.max = QVariant(50);
   secondYtiles.group = "second tiles";
   secondYtiles.id = "secondYtiles";
   configurables.append(secondYtiles);
}
void TransformTextureGenerator::generate(QSize size, TexturePixel* destimage,
                                         const QMap<QString, TextureImagePtr>& sourceimages,
                                         const TextureNodeSettings& settings) const {
   if (!destimage || !size.isValid()) {
      return;
   }

   double scaleX = settings.value("xscale").toDouble() / 100;
   double scaleY = settings.value("yscale").toDouble() / 100;
   double rotation = settings.value("rotation").toDouble();
   int offsetLeft = settings.value("offsetleft").toDouble() * size.width() / 100;
   int offsetTop = settings.value("offsettop").toDouble() * size.height() / 100;
   int firstXtiles = settings.value("firstXtiles").toInt();
   int firstYtiles = settings.value("firstYtiles").toInt();
   int secondXtiles = settings.value("secondXtiles").toInt();
   int secondYtiles = settings.value("secondYtiles").toInt();
   QColor backgroundcolor = settings.value("backgroundcolor").value<QColor>();

   QImage tempimage = makeTextureImageView(size, destimage);

   if (sourceimages.contains(QStringLiteral("Image"))) {
      memcpy(destimage, sourceimages.value(QStringLiteral("Image"))->getData(),
             size.width() * size.height() * sizeof(TexturePixel));
   } else {
      memset(destimage, 0, size.width() * size.height() * sizeof(TexturePixel));
   }

   TextureImage tiledStorage(size);
   QImage tiledimage = tiledStorage.toQImageView();
   {
      QPainter firstpainter(&tiledimage);
      firstpainter.scale(static_cast<double>(1) / secondXtiles,
                         static_cast<double>(1) / secondYtiles);
      firstpainter.fillRect(0, 0, secondXtiles * tempimage.width(),
                            secondYtiles * tempimage.height(), QBrush(tempimage));
   }

   TextureImage destinationStorage(QSize(firstXtiles * size.width(), firstYtiles * size.height()));
   QImage destBuffer = destinationStorage.toQImageView();
   TexturePixel col(backgroundcolor.red(), backgroundcolor.green(), backgroundcolor.blue(),
                    backgroundcolor.alpha());
   for (std::size_t index = 0; index < destinationStorage.pixelCount(); ++index) {
      destinationStorage.data()[index] = col;
   }

   {
      QPainter painter(&destBuffer);
      painter.translate(offsetLeft, offsetTop);
      painter.translate(size.width() / 2, size.height() / 2);
      painter.rotate(rotation);
      painter.scale(scaleX, scaleY);
      painter.translate(-firstXtiles * size.width() / 2, -firstYtiles * size.height() / 2);
      painter.fillRect(0, 0, destBuffer.width(), destBuffer.height(), QBrush(tiledimage));
   }

   memset(destimage, 0, size.width() * size.height() * sizeof(TexturePixel));
   QPainter finalpainter(&tempimage);
   finalpainter.drawImage(QRectF(0, 0, size.width(), size.height()), destBuffer,
                          QRectF(0, 0, size.width(), size.height()));
}

File: generators/transform.js
const generator = {
  apiVersion: 1,
  name: "Transform",
  description: "Scales, rotates, offsets, and tiles the input image.",
  type: "filter",
  inputs: ["Image"],

  // Each entry creates a control in the node settings panel.
  settings: [
    {
      id: "xscale",
      type: "real",
      name: "Horizontal scale (%)",
      description: "Scales the input image horizontally around the texture centre.",
      default: 100,
      min: 0,
      max: 500,
      group: "scale",
    },
    {
      id: "yscale",
      type: "real",
      name: "Vertical scale (%)",
      description: "Scales the input image vertically around the texture centre.",
      default: 100,
      min: 0,
      max: 500,
      group: "scale",
    },
    {
      id: "rotation",
      type: "real",
      name: "Rotation (°)",
      description: "Rotates the input image around the texture centre.",
      default: 0,
      min: -360,
      max: 360,
    },
    {
      id: "offsetleft",
      type: "real",
      name: "Horizontal offset (%)",
      description: "Moves the input image horizontally across the texture.",
      default: 0,
      min: -200,
      max: 200,
    },
    {
      id: "offsettop",
      type: "real",
      name: "Vertical offset (%)",
      description: "Moves the input image vertically across the texture.",
      default: 0,
      min: -200,
      max: 200,
    },
    {
      id: "backgroundcolor",
      type: "color",
      name: "Background colour",
      description: "Colour used where the transformed image leaves empty space.",
      default: { r: 0, g: 0, b: 0, a: 0 },
    },
    {
      id: "firstXtiles",
      type: "integer",
      name: "First-pass horizontal tiles",
      description: "Number of source-sized tiles across the area transformed as one image.",
      default: 1,
      min: 1,
      max: 5,
      group: "first tiles",
    },
    {
      id: "firstYtiles",
      type: "integer",
      name: "First-pass vertical tiles",
      description: "Number of source-sized tiles down the area transformed as one image.",
      default: 1,
      min: 1,
      max: 5,
      group: "first tiles",
    },
    {
      id: "secondXtiles",
      type: "integer",
      name: "Second-pass horizontal tiles",
      description: "Number of source copies fitted horizontally within each tile.",
      default: 1,
      min: 1,
      max: 50,
      group: "second tiles",
    },
    {
      id: "secondYtiles",
      type: "integer",
      name: "Second-pass vertical tiles",
      description: "Number of source copies fitted vertically within each tile.",
      default: 1,
      min: 1,
      max: 50,
      group: "second tiles",
    },
  ],

  // This function runs whenever the application renders the node.
  generate(size, settings, output, inputs) {
    // Step 1: fill the output with the selected background colour. Parts of the
    // transformed image that do not cover the output will leave this colour visible.
    const outputPixels = output.data;
    const width = size.width;
    const height = size.height;
    const backgroundColour = settings.backgroundcolor;

    // Write one RGBA pixel first. outputPixels.fill() cannot be used because it
    // repeats one number rather than a four-byte colour.
    outputPixels[0] = backgroundColour.r;
    outputPixels[1] = backgroundColour.g;
    outputPixels[2] = backgroundColour.b;
    outputPixels[3] = backgroundColour.a;

    // Repeatedly copy the completed bytes over the empty part. The filled area
    // doubles each time, which is much faster than writing every pixel in JavaScript.
    for (let filled = 4; filled < outputPixels.length; filled *= 2) {
      const copyLength = Math.min(filled, outputPixels.length - filled);
      outputPixels.set(outputPixels.subarray(0, copyLength), filled);
    }

    // Step 2: read the source image and convert percentages into scale factors.
    // A scale of 100% becomes 1, 50% becomes 0.5, and 200% becomes 2.
    const sourceImage = inputs.Image;
    const horizontalScale = settings.xscale / 100;
    const verticalScale = settings.yscale / 100;

    // There is nothing to transform without an image or with an invisible scale.
    if (!sourceImage || horizontalScale <= 0 || verticalScale <= 0) return;

    const sourcePixels = sourceImage.data;
    // A stride is the number of bytes from the start of one row to the next.
    const sourceStride = sourceImage.stride;
    const outputStride = output.stride;

    // Step 3: describe the virtual tiled source that will be transformed.
    // The first-pass settings enlarge this area to contain several source-sized tiles.
    const tiledSourceWidth = settings.firstXtiles * width;
    const tiledSourceHeight = settings.firstYtiles * height;

    // The second-pass settings repeat the source within every source-sized tile.
    // For example, a value of 2 fits two complete copies across one tile.
    const horizontalRepeats = settings.secondXtiles;
    const verticalRepeats = settings.secondYtiles;

    // The destination centre stays fixed while the image rotates. Keeping its
    // half-pixel part is important for odd-sized textures and tiled source areas.
    const transformedCenterX = width / 2 + settings.offsetleft * width / 100;
    const transformedCenterY = height / 2 + settings.offsettop * height / 100;

    // JavaScript's trigonometric functions use radians rather than degrees.
    const radians = settings.rotation * Math.PI / 180;
    const cosine = Math.cos(radians);
    const sine = Math.sin(radians);

    // Step 4: calculate the output rectangle that can contain the transformed image.
    // Rotating a rectangle makes its horizontal and vertical reach depend on both
    // original dimensions. These extents give its furthest distance from the centre.
    const halfScaledWidth = tiledSourceWidth * horizontalScale / 2;
    const halfScaledHeight = tiledSourceHeight * verticalScale / 2;
    const extentX = Math.abs(cosine) * halfScaledWidth + Math.abs(sine) * halfScaledHeight;
    const extentY = Math.abs(sine) * halfScaledWidth + Math.abs(cosine) * halfScaledHeight;
    const left = Math.max(0, Math.floor(transformedCenterX - extentX));
    const right = Math.min(width - 1, Math.ceil(transformedCenterX + extentX));
    const top = Math.max(0, Math.floor(transformedCenterY - extentY));
    const bottom = Math.min(height - 1, Math.ceil(transformedCenterY + extentY));

    // Step 5: visit every output pixel in that rectangle. Instead of pushing source
    // pixels forwards, work backwards: undo the transformation to discover which
    // source pixel belongs at each output position. This avoids gaps in the result.
    const virtualXChangePerPixel = cosine / horizontalScale;
    const virtualYChangePerPixel = -sine / verticalScale;

    for (let y = top; y <= bottom; ++y) {
      const xFromCenter = left + 0.5 - transformedCenterX;
      const yFromCenter = y + 0.5 - transformedCenterY;

      // Undo rotation and scaling, then move the origin from the centre to the
      // top-left corner of the virtual tiled source.
      let virtualSourceX = (cosine * xFromCenter + sine * yFromCenter)
          / horizontalScale
        + tiledSourceWidth / 2;
      let virtualSourceY = (-sine * xFromCenter + cosine * yFromCenter)
          / verticalScale
        + tiledSourceHeight / 2;
      let outputOffset = y * outputStride + left * 4;

      for (let x = left; x <= right; ++x) {
        const insideTiledSource = virtualSourceX >= 0
          && virtualSourceX < tiledSourceWidth
          && virtualSourceY >= 0
          && virtualSourceY < tiledSourceHeight;

        if (insideTiledSource) {
          // Multiplication creates the second-pass repetitions. The remainder
          // operator (%) then wraps the coordinate into the original source image.
          const repeatedSourceX = Math.floor(virtualSourceX * horizontalRepeats);
          const repeatedSourceY = Math.floor(virtualSourceY * verticalRepeats);
          const sourceX = repeatedSourceX % width;
          const sourceY = repeatedSourceY % height;
          const sourceOffset = sourceY * sourceStride + sourceX * 4;
          const sourceAlpha = sourcePixels[sourceOffset + 3] / 255;

          // Step 6: copy opaque pixels directly. They completely hide the background.
          if (sourceAlpha >= 1) {
            outputPixels[outputOffset] = sourcePixels[sourceOffset];
            outputPixels[outputOffset + 1] = sourcePixels[sourceOffset + 1];
            outputPixels[outputOffset + 2] = sourcePixels[sourceOffset + 2];
            outputPixels[outputOffset + 3] = sourcePixels[sourceOffset + 3];
          } else if (sourceAlpha > 0) {
            // A transparent background has no existing colour to preserve.
            const backgroundAlpha = outputPixels[outputOffset + 3] / 255;
            if (backgroundAlpha === 0) {
              outputPixels[outputOffset] = sourcePixels[sourceOffset];
              outputPixels[outputOffset + 1] = sourcePixels[sourceOffset + 1];
              outputPixels[outputOffset + 2] = sourcePixels[sourceOffset + 2];
              outputPixels[outputOffset + 3] = sourcePixels[sourceOffset + 3];
            } else {
              // For two translucent colours, retain the visible contribution from
              // both. The result is stored as ordinary, non-premultiplied RGBA.
              const remainingBackground = 1 - sourceAlpha;
              const resultAlpha = sourceAlpha + backgroundAlpha * remainingBackground;
              outputPixels[outputOffset] = Math.round(
                (sourcePixels[sourceOffset] * sourceAlpha
                  + outputPixels[outputOffset] * backgroundAlpha * remainingBackground)
                  / resultAlpha,
              );
              outputPixels[outputOffset + 1] = Math.round(
                (sourcePixels[sourceOffset + 1] * sourceAlpha
                  + outputPixels[outputOffset + 1] * backgroundAlpha * remainingBackground)
                  / resultAlpha,
              );
              outputPixels[outputOffset + 2] = Math.round(
                (sourcePixels[sourceOffset + 2] * sourceAlpha
                  + outputPixels[outputOffset + 2] * backgroundAlpha * remainingBackground)
                  / resultAlpha,
              );
              outputPixels[outputOffset + 3] = Math.round(resultAlpha * 255);
            }
          }
        }

        // Move to the next output pixel. Four bytes advance to its RGBA values.
        virtualSourceX += virtualXChangePerPixel;
        virtualSourceY += virtualYChangePerPixel;
        outputOffset += 4;
      }
    }
  },
};
